node_exporter.go

```diff
@@ -28,80 +28,139 @@ import (
+// created on the fly, if filtering is requested. Create instances with
+// newHandler.
+type handler struct {
+	unfilteredHandler http.Handler
+	// exporterMetricsRegistry is a separate registry for the metrics about
+	// the exporter itself.
+	exporterMetricsRegistry *prometheus.Registry
+	includeExporterMetrics  bool
+	maxRequests             int
 }
 
-func handler(w http.ResponseWriter, r *http.Request) {
+func newHandler(includeExporterMetrics bool, maxRequests int) *handler {
+	h := &handler{
+		exporterMetricsRegistry: prometheus.NewRegistry(),
+		includeExporterMetrics:  includeExporterMetrics,
+		maxRequests:             maxRequests,
+	}
+	if h.includeExporterMetrics {
+		h.exporterMetricsRegistry.MustRegister(
+			prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
+			prometheus.NewGoCollector(),
+		)
+	}
+	if innerHandler, err := h.innerHandler(); err != nil {
+		log.Fatalf("Couldn't create metrics handler: %s", err)
+	} else {
+		h.unfilteredHandler = innerHandler
+	}
+	return h
+}
+
+// ServeHTTP implements http.Handler.
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	filters := r.URL.Query()["collect[]"]
 	log.Debugln("collect query:", filters)
 
-	nc, err := collector.NewNodeCollector(filters...)
+	if len(filters) == 0 {
+		// No filters, use the prepared unfiltered handler.
+		h.unfilteredHandler.ServeHTTP(w, r)
+		return
+	}
+	// To serve filtered metrics, we create a filtering handler on the fly.
+	filteredHandler, err := h.innerHandler(filters...)
 	if err != nil {
-		log.Warnln("Couldn't create", err)
+		log.Warnln("Couldn't create filtered metrics handler:", err)
 		w.WriteHeader(http.StatusBadRequest)
-		w.Write([]byte(fmt.Sprintf("Couldn't create %s", err)))
+		w.Write([]byte(fmt.Sprintf("Couldn't create filtered metrics handler: %s", err)))
 		return
 	}
+	filteredHandler.ServeHTTP(w, r)
+}
 
-	registry := prometheus.NewRegistry()
-	err = registry.Register(nc)
+// innerHandler is used to create buth the one unfiltered http.Handler to be
+// wrapped by the outer handler and also the filtered handlers created on the
+// fly. The former is accomplished by calling innerHandler without any arguments
+// (in which case it will log all the collectors enabled via command-line
+// flags).
+func (h *handler) innerHandler(filters ...string) (http.Handler, error) {
+	nc, err := collector.NewNodeCollector(filters...)
 	if err != nil {
-		log.Errorln("Couldn't register collector:", err)
-		w.WriteHeader(http.StatusInternalServerError)
-		w.Write([]byte(fmt.Sprintf("Couldn't register collector: %s", err)))
-		return
+		return nil, fmt.Errorf("couldn't create collector: %s", err)
+	}
+
+	// Only log the creation of an unfiltered handler, which should happen
+	// only once upon startup.
+	if len(filters) == 0 {
+		log.Infof("Enabled collectors:")
+		collectors := []string{}
+		for n := range nc.Collectors {
+			collectors = append(collectors, n)
+		}
+		sort.Strings(collectors)
```

Comment:
Why do we sort collectors?

Reply:
To avoid extra Prometheus reload if we just refactored configuration generator without changing the meaning.
We do the same for of Agents there (`ORDER BY agent_type, agent_id`) and exporter flags.

Reply:
Ah, doh :) We do that in pmm-managed.

There we do it only because upstream is doing it: https://github.com/percona/node_exporter/pull/15/files/f78b76e77d8943e3a01f5d3c4d0004e7d564b60b#r319979026